### ISSUE TITLE
Add ANTS (ITK) linear transform conversion to transformcalc

### ISSUE THREAD

**Issue by draffelt** [MRtrix3/mrtrix3#602]

**Comment by draffelt** [comment]
I have code to do this (in the old syntax). I was just about to add it and saw all the recent features in `transformcalc`. 

I was thinking that perhaps we should split out the FSL import to a separate command (`transformconvert`), with the ITK conversion. What do you think?

Note we already have `warpconvert`. So not sure that name `transformconvert` is perfect...since a warp is a transformation.

**Comment by thijsdhollander** [comment]
I like the idea of a command like `transformconvert`.  As to name... hmm, the only more specific name I can think of is `tranformationmatrixconvert`. :grin:   It's pretty hard to make this one more specific (to distinguish from `warpconvert`) and not end up with crazy long names.

**Comment by draffelt** [comment]
If we go with transformconvert, we can just make sure it's clear in the
help what it does, and also refer the user to warpconvert for non-linear
transforms.

**Comment by thijsdhollander** [comment]
Yep, I agree.  Unless anyone else has brilliant ideas on a name, I think we're not going to come up with anything more elegant than just `transformconvert`.

**Comment by jdtournier** [comment]
I'm sure we've had this discussion before...  

The only other name I can think of is `affineconvert` - but that would be purely to disambiguate its purpose, not sure I actually prefer that over `transformconvert`... 

I reckon as long as the documentation is clear, it's not that big a deal. Go for `transformconvert`.

**Comment by maxpietsch** [comment]
I would keep it consistent with `transformclac` as both deal with the same type of transformation. I see that `warp` and `transform` can easily be confused but I think `affine` instead of `transform` is confusing as well in the case of rigid registration. 
How about `lintransformconvert` and `lintransformcalc`?

**Comment by draffelt** [comment]
`lintransformconvert` is not bad, but probably less discoverable.

**Comment by jdtournier** [comment]
I think as long as we're consistent, it shouldn't matter a great deal. If there's already a few `transform` commands, let's stick with that - users will quickly cotton on to the fact that we use the term `transform` for affine matrices, and `warp` for non-linear deformations... I think it's clear enough.

**Comment by maxpietsch** [comment]
Happy with `transformconvert`. Might be worth changing the description of `mrregister`'s `-nl_warp_full` which talks about "linear and non-linear transform" to "linear transformation and non-linear warps".

@draffelt I am happy to split `transformcalc` if you give me the old ITK import code.

**Comment by draffelt** [comment]
Cool. Thanks Max. If you are offering I'm happy! It's in the transformcalc
command in the registration branch on Gitlab.  Unfortunately it's in the
old syntax, but not too many lines of code.

**Comment by maxpietsch** [comment]
Ok, had a go at it. I can't quite get the testing data repo to load the data I added. 

    mrtrix3/testing/data [HEAD]$ git branch -a
    * (detached from 99cefe0)
      master
      remotes/origin/HEAD -> origin/master
      remotes/origin/master

Last git log entry is 99cefe043c0f2ab6b91abf9. I was hoping to see [47f6a1f0c75ca75ab9c2](https://github.com/MRtrix3/test_data/commit/47f6a1f0c75ca75ab9c266672ca814f14d96d7d1).

**Comment by jdtournier** [comment]
Did you [update the main repo](https://github.com/MRtrix3/mrtrix3/blob/master/testing/README.md#updating-the-main-mrtrix3-repo-to-make-use-of-the-latest-commits) as per the readme...?

**Comment by jdtournier** [comment]
Although having just gone through the process myself, I noticed that I had to also `git add testing/data` before committing....

**Comment by maxpietsch** [comment]
I think I followed the readme to the letter. Now after `git add testing/data` in the root folder of this branch, my test data is there but tckmap and labelconfig are missing. They are there though if I run the tests from master.

**Comment by jdtournier** [comment]
When you say they're not there, what do you mean exactly? There's been a few changes to the test data for some of the tests in anticipation of the pending `tag_0.3.15` merge. It does cause a few tests to fail, but shouldn't matter once merged...

**Comment by maxpietsch** [comment]
Um, with not there I mean that there is no folder `tckmap` or `labelconfig` in testing/data. I find this git submodule thing a bit confusing. I guess we can just see if testing breaks after merge with master but here is the output for a fresh clone in case it is worth figuring out what is going on:

    git clone https://github.com/MRtrix3/mrtrix3.git mrtrix3_testing
    cd mrtrix3_testing
    git checkout registration_transformconvert
    ./configure && ./build
    ./run_tests

Test data:

    ~/mrtrix3_tests/testing/data [HEAD]$ git log -2
    commit 47f6a1f0c75ca75ab9c266672ca814f14d96d7d1
    Author: Max Pietsch <mail@maxpietsch.com>
    Date:   Fri Jun 10 19:23:28 2016 +0100
    
        added transformconvert itk testdata generated with transformcalc 63464b95374aa8bc3be37341d002
    
    commit b956138f3bec7a3e2af2198425055513c8ac728e
    Author: J-Donald Tournier <jdtournier@gmail.com>
    Date:   Fri Jun 10 18:20:56 2016 +0100
    
        add data for dwidenoise

testing.log excerpt:

    ## running "label2colour"...
    
    # command: label2colour labelconfig/in_fs.mif.gz -lut_freesurfer labelconfig/FreeSurferColorLUT.txt - | testing_diff_data - label2colour/out_fs_all.mif.gz 0.5 [ ERROR ]
    label2colour: [ERROR] input file "labelconfig/FreeSurferColorLUT.txt" not found (required for option "-lut_freesurfer")
    testing_diff_data: [ERROR] no filename supplied to standard input (broken pipe?)
    testing_diff_data: [ERROR] error opening image "-"                                                                                                                        
    
    # command: label2colour labelconfig/out_fs.mif.gz -lut_freesurfer labelconfig/FreeSurferColorLUT.txt -config labelconfig/fs_default.txt - | testing_diff_data - label2colour/    out_fs_config.mif.gz 0.5 [ ERROR ]
    label2colour: [ERROR] input file "labelconfig/FreeSurferColorLUT.txt" not found (required for option "-lut_freesurfer")
    testing_diff_data: [ERROR] no filename supplied to standard input (broken pipe?)
    testing_diff_data: [ERROR] error opening image "-"
    
    # command: label2colour labelconfig/in_iit.mif.gz -lut_itksnap labelconfig/LUT_GM.txt - | testing_diff_data - label2colour/out_iit.mif.gz 0.5 [ ERROR ]
    label2colour: [ERROR] input file "labelconfig/LUT_GM.txt" not found (required for option "-lut_itksnap")
    testing_diff_data: [ERROR] no filename supplied to standard input (broken pipe?)
    testing_diff_data: [ERROR] error opening image "-"
    
    
    ## ERROR: 3 tests failed for "label2colour"

### master

All tests pass if I checkout the master branch, ./build and rerun run_tests.
After that I can see  label2colour/ label2mesh/   labelconfig/ in testing/data

and

    :~/mrtrix3_tests/testing/data [HEAD]$ git log -2   
    commit 99cefe043c0f2ab6b91abf9858a153e061f8a0bd
    Author: Ben Jeurissen <ben.jeurissen@gmail.com>
    Date:   Mon May 16 18:45:22 2016 +0200
    
        dwi2tensor test data
    
    commit b4a0263980d89eb2f698ebc7b89c255705f73301
    Author: Dave Raffelt <draffelt@gmail.com>
    Date:   Fri May 6 14:04:17 2016 +1000
    
        subtle changed to some output files due to change in default smoothing extent

**Comment by jdtournier** [comment]
OK, just looked into this: as expected, this is entirely down to changes in `tag_0.3.15`:
~~~ShellSession
$ git log -p tag_0.3.15 -- tests/label2colour
commit f7502ff8d0d2127744fe0d5aad3f21421c00839f
Author: Robert Smith <r.smith@brain.org.au>
Date:   Mon Jun 6 16:20:35 2016 +1000

    Various fixes preparing for 0.3.15 tag

diff --git a/testing/tests/label2colour b/testing/tests/label2colour
index 2b66fd3..f43fb1b 100644
--- a/testing/tests/label2colour
+++ b/testing/tests/label2colour
@@ -1,3 +1,2 @@
-label2colour labelconfig/in_fs.mif.gz -lut_freesurfer labelconfig/FreeSurferColorLUT.txt - | testing_diff_data - label2colour/out_fs_all.mif.gz 0.5
-label2colour labelconfig/out_fs.mif.gz -lut_freesurfer labelconfig/FreeSurferColorLUT.txt -config labelconfig/fs_default.txt - | testing_diff_data - label2colour/out_fs_config.mif.gz 0.5
-label2colour labelconfig/in_iit.mif.gz -lut_itksnap labelconfig/LUT_GM.txt - | testing_diff_data - label2colour/out_iit.mif.gz 0.5
+label2colour labelconvert/in_fs.mif.gz -lut labelconvert/FreeSurferColorLUT.txt - | testing_diff_data - label2colour/out_fs.mif.gz 0.5
+label2colour labelconvert/in_iit.mif.gz -lut labelconvert/LUT_GM.txt - | testing_diff_data - label2colour/out_iit.mif.gz 0.5
~~~

I do agree that the submodule thing is confusing, but I can't actually think of a simpler way of managing this without having the test data actually reside in the same repo as the code... One thing that might make things easier would be to try to maintain the same branching structure on the `test_data` repo as on the code repo, so that the `master` branches remains in sync. That would require a little bit more fiddling for developers that want to _modify_ the test data in ways that break existing tests, but given that this should be a relatively rare case, and only done by experienced devs anyway, maybe that would be the better way to do things...

Basically, what you'd like to happen is that you checkout the `master` branch on the `test_data` repo, push your new test data, and everything would work (assuming your code branches from `master` on the code repo). The problem is when people make changes on the `test_data` repo that would break things for `master`, and then someone else branches off that to add their own tests, everything falls apart. Thankfully, the current `master` keeps checking out fine in testing since it's requesting the specific SHA1 it was set to use at the time. 

So one way around this would be to push breaking changes to `test_data` into a seperate branch, ideally named identically to the branch it's being designed for on the code repo, and create an explicit pull request on that side too to mirror the one on the master branch. That way other devs can push their own test data onto the `master` branch and not have testing fail unexpectedly. Once the branch with the breacking changes is ready to merge, there'd be a bit of housekeeping to do to make sure things go smoothly:

- on `test_data`: merge `master` into your breaking branch

- on the main repo: record updated SHA1 of current `test_data` repo if needed, and merge pull request normally

- back on `test_data`: do _fast-forward_ merge of your pull request. This is not the default on GitHUb, normlly it would add an explicit merge commit even when a straight fast-forward is possible, since this helps keep track of the history. But since it creates an additional commit, this would mean the SHA1 used in the main repo would not actually be the `master` branch - although I'm not sure whether that would make any difference in practice since the data itself would be identical for both commits....

So a bit of a faff to do this properly, but note this would _only_ be for changes that modify the test data. Adding more data is fine. So if you really need to change the data on the `test_data` repo, either it's a quick fix and you merge your pull request on the code repo very quickly (i.e. you've just fixed a bug and need to update the test data accordingly), or push the changes to a separate branch. You'd then need to figure out the right `git submodule` call to make sure you record the correct SHA1 commit of the `test_data` module into your branch on the main repo. 

All very simple, right...? :wink:

**Comment by draffelt** [comment]
I guess we can close this off? And make a new issue to discuss the testing submodule?
